### PR TITLE
ocamlPackages.tezos-*: 8.3 -> 10.2

### PR DESCRIPTION
**Pull request by ulrikstrid**
###### Motivation for this change

This updates all the tezos-packages to the latest release in preparation of the updated LIGO version that is currently being reviewed.

This PR includes some new needed packages to build tezos.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/ocaml-modules/getopt/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  buildInputs = [
+    ocaml
+    findlib
+    ocamlbuild
+  ];
+
+  doCheck = true;
+  createFindlibDestdir = true;
+
+  meta = {
```
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/getopt/default.nix:21]: This is missing a `platforms` field.
```suggestion
  meta = {
    inherit (ocaml.meta) platforms;
```

File: pkgs/development/ocaml-modules/getopt/default.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
+{ lib, fetchzip, stdenv, ocaml, findlib, ocamlbuild }:
+
+stdenv.mkDerivation {
+  pname = "getopt";
+  version = "20120615";
+
+  src = fetchzip {
+    url = https://download.ocamlcore.org/ocaml-getopt/ocaml-getopt/20120615/ocaml-getopt-20120615.tar.gz;
```
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/getopt/default.nix:8]:
```suggestion
stdenv.mkDerivation rec {
  pname = "getopt";
  version = "20120615";

  src = fetchzip {
    url = https://download.ocamlcore.org/ocaml-getopt/ocaml-getopt/${version}/ocaml-getopt-${version}.tar.gz;
```

File: pkgs/development/ocaml-modules/hashcons/default.nix [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  src = fetchFromGitHub {
+    owner  = "backtracking";
+    repo   = "ocaml-${pname}";
+    rev    = "d733325eeb55878bed285120c2c088daf78f0e2b";
+    sha256 = "0h4pvwj34pndaw3pajkhl710ywwinhc9pqimgllfmkl37wz2d8zq";
+  };
+
+  useDune2 = true;
+
+  postPatch = "ls";
```
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/hashcons/default.nix:16]: This can be removed?

Comment (ulrikstrid) [reply pkgs/development/ocaml-modules/hashcons/default.nix:16]: Yes, sorry

**Comment by ulrikstrid** [comment]
Thank you @vbgl for the fixes, I have committed and squashed them into the getopt commit

**Review by vbgl — COMMENTED** [review]
Please squash together the commits that are not independent. It will make reviewing simpler.

**Comment by vbgl** [comment]
Yes. For instance, updating json-data-encoding breaks some libraries. Are these breakages fixed by other commits in your PR? I don’t know (and I prefer not to have to figure this out by myself).

**Comment by vbgl** [comment]
@GrahamcOfBorg build ocaml-ng.ocamlPackages_4_10.tezos-protocol-demo-noops

**Pull request merged**